### PR TITLE
GH-3927 optimize query optimizers

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #3927 

Briefly describe the changes proposed in this PR:

 - Added [AbstractSimpleQueryModelVisitor.java](https://github.com/eclipse/rdf4j/blob/c8dcaef15b26b815125906debaa778aa1884fbe3/core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/AbstractSimpleQueryModelVisitor.java) which is a leaner and faster implementation of the QueryModelVisitor
 - Copied over all the regular optimizers into a new package and changed them to use the new AbstractSimpleQueryModelVisitor
 - Optimized the join optimizer
 - Optimized the query algebra
 - Deprecated the old optimizers

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

**Comment by hmottestad** [comment]
## Before

```
```

## After
```
Benchmark                                    Mode  Cnt   Score   Error  Units
ParallelQueryBenchmark.mixedQueriesAndReads  avgt    5  35.377 ± 1.567  ms/op
```

**Comment by jeenbroekstra** [comment]
that's a lot of wip

**Comment by hmottestad** [comment]
Indeed. I'm extracting stuff from my big performance improvements branch and it's a lot of back and forth trying to carve out the right bits and still have all the test pass and japicmp not complain too much. 

Performance numbers are looking good though. An overall performance boost for the MemoryStore and a 2x improvement for concurrent queries.

**Comment by hmottestad** [comment]
For comparison. The current state of the main branch:

```
Benchmark                                    Mode  Cnt   Score    Error  Units
ParallelQueryBenchmark.mixedQueriesAndReads  avgt    5  86.217 ± 10.537  ms/op

Benchmark                                                     Mode  Cnt      Score     Error  Units
QueryBenchmark.complexQuery                                   avgt    5      4.815 ±   0.079  ms/op
QueryBenchmark.different_datasets_with_similar_distributions  avgt    5      2.412 ±   0.070  ms/op
QueryBenchmark.groupByQuery                                   avgt    5      1.220 ±   0.117  ms/op
QueryBenchmark.long_chain                                     avgt    5    713.659 ±  19.454  ms/op
QueryBenchmark.lots_of_optional                               avgt    5    202.711 ±   5.901  ms/op
QueryBenchmark.minus                                          avgt    5  15092.775 ± 274.211  ms/op
QueryBenchmark.nested_optionals                               avgt    5    210.971 ±   6.627  ms/op
QueryBenchmark.pathExpressionQuery1                           avgt    5     19.107 ±   0.706  ms/op
QueryBenchmark.pathExpressionQuery2                           avgt    5      1.308 ±   0.087  ms/op
QueryBenchmark.query_distinct_predicates                      avgt    5     88.170 ±   3.452  ms/op
QueryBenchmark.simple_filter_not                              avgt    5     10.155 ±   0.581  ms/op
```

**Comment by hmottestad** [comment]
So there is a bunch of stuff in this branch now. I'll have to at least split things into separate commits. 

Comparing performance against the current state of the main branch is pretty cool. With all the performance improvements currently on develop together with the improvements in this branch the overall query performance is often around 2x better and more importantly a bit over 2x for concurrent queries.

I have long intended to create some reusable benchmarks so we can compare the NativeStore, LMDB Store and the MemoryStore. The overhead of disk access will probably reduce the performance impact, but there should still be some improvements. 

I also want to post some ShaclSail benchmarks, since the validation typically generates a lot of queries many of which take more time to optimize than to evaluate.

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/optimizer/FilterOptimizer.java [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+ *  </code>
+ *
+ * This optimization allows for sharing evaluation costs in the future and removes an iterator. This is done as a second
+ * step to not break the first optimization. In the case that the splitting was done but did not help it is now undone.
+ *
+ * @author Arjohn Kampman
+ * @author Jerven Bolleman
+ */
+public class FilterOptimizer implements QueryOptimizer {
+
```
Comment (hmottestad) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/optimizer/FilterOptimizer.java:76]: I've kept the original authors, but I have updated the copyright.

**Comment by hmottestad** [comment]
The benchmarks aren't accurate anymore since I split out the remainder of this PR into a separate branch: https://github.com/eclipse/rdf4j/pull/3943

**Review by jeenbroekstra — APPROVED** [review]

**Pull request merged**